File: src/entries.rs
use std::fs;
use std::path::{Path, PathBuf};

#[derive(Clone, Debug)]
pub struct GameEntry {
    pub id: u32,
    pub name: String,
    pub exec: String,
    pub cover: Option<String>,
}

#[derive(serde::Deserialize)]
struct TomlGame {
    name: Option<String>,
    exec: Option<String>,
    cover: Option<String>,
}

pub fn load_games(dirs: &[PathBuf]) -> Vec<GameEntry> {
    let mut games = Vec::new();
    let mut next_id: u32 = 0;

    for dir in dirs {
        let Ok(read_dir) = fs::read_dir(dir) else {
            continue;
        };

        let mut paths = Vec::new();
        for entry in read_dir.flatten() {
            let path = entry.path();
            if path.is_file() && path.extension().and_then(|s| s.to_str()) == Some("toml") {
                paths.push(path);
            }
        }
        paths.sort();

        for path in paths {
            if let Some(mut game) = parse_game(&path) {
                if !game.exec.is_empty() {
                    game.id = next_id;
                    next_id += 1;
                    games.push(game);
                }
            }
        }
    }

    if games.is_empty() {
        games.push(GameEntry {
            id: 0,
            name: "Howdy!".to_string(),
            exec: String::new(),
            cover: None,
        });
    }

    games
}

pub fn parse_game(path: &Path) -> Option<GameEntry> {
    let contents = fs::read_to_string(path).ok()?;
    parse_game_contents(&contents)
}

pub fn parse_game_contents(contents: &str) -> Option<GameEntry> {
    // Try standard TOML parsing first
    if let Ok(tg) = toml::from_str::<TomlGame>(contents) {
        if let (Some(name), Some(exec)) = (tg.name, tg.exec) {
            return Some(GameEntry {
                id: 0,
                name,
                exec,
                cover: tg.cover,
            });
        }
    }

    // Fallback to manual line-by-line parsing (matching the original C++ behavior)
    let mut name: Option<String> = None;
    let mut exec: Option<String> = None;
    let mut cover: Option<String> = None;

    for raw_line in contents.lines() {
        let line = raw_line.trim();
        if line.is_empty() || line.starts_with('#') || line.starts_with('[') {
            continue;
        }

        let Some((key_part, val_part)) = line.split_once('=') else {
            continue;
        };

        let key = key_part.trim();
        let mut value = val_part.trim();
        if value.starts_with('"') && value.ends_with('"') && value.len() >= 2 {
            value = &value[1..value.len() - 1];
        }

        match key {
            "name" => name = Some(value.to_string()),
            "exec" => exec = Some(value.to_string()),
            "cover" => cover = Some(value.to_string()),
            _ => {}
        }
    }

    let name = name?;
    let exec = exec?;

    Some(GameEntry {
        id: 0,
        name,
        exec,
        cover,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_standard_toml() {
        let toml_str = r#"
            name = "Sonic Adventure"
            exec = "sonic-adventure"
            cover = "/games/sonic/cover.png"
        "#;
        let game = parse_game_contents(toml_str).unwrap();
        assert_eq!(game.name, "Sonic Adventure");
        assert_eq!(game.exec, "sonic-adventure");
        assert_eq!(game.cover, Some("/games/sonic/cover.png".to_string()));
    }

    #[test]
    fn test_parse_unquoted() {
        let toml_str = "name = Sonic\nexec = sonic\n";
        let game = parse_game_contents(toml_str).unwrap();
        assert_eq!(game.name, "Sonic");
        assert_eq!(game.exec, "sonic");
        assert_eq!(game.cover, None);
    }
}

File: src/entries.cpp
#include "entries.h"

#include <fstream>
#include <iterator>
#include <sstream>

namespace fs = std::filesystem;

std::vector<GameEntry> load_games(const std::vector<fs::path>& dirs) {
    std::vector<GameEntry> games;
    uint32_t next_id = 0;

    for (const auto& dir : dirs) {
        std::error_code ec;
        for (const auto& entry : fs::directory_iterator(dir, ec)) {
            if (entry.is_regular_file() && entry.path().extension() == ".toml") {
                auto game = parse_game(entry.path());
                if (game && !game->exec.empty()) {
                    game->id = next_id++;
                    games.push_back(std::move(*game));
                }
            }
        }
    }

    if (games.empty()) {
        games.push_back(GameEntry{
            .id = 0,
            .name = "Howdy!",
            .exec = "",
            .cover = std::nullopt,
        });
    }

    return games;
}

std::optional<GameEntry> parse_game(const fs::path& path) {
    std::ifstream file;
    file.open(path.c_str());
    if (!file.is_open()) {
        return std::nullopt;
    }

    std::string contents((std::istreambuf_iterator<char>(file)),
                          std::istreambuf_iterator<char>());
    return parse_game_contents(contents);
}

std::optional<GameEntry> parse_game_contents(const std::string& contents) {
    std::optional<std::string> name;
    std::optional<std::string> exec;
    std::optional<std::string> cover;

    std::istringstream stream(contents);
    std::string line;

    while (std::getline(stream, line)) {
        auto start = line.find_first_not_of(" \t\r\n");
        if (start == std::string::npos) continue;
        auto end = line.find_last_not_of(" \t\r\n");
        line = line.substr(start, end - start + 1);

        if (line.empty() || line[0] == '#') continue;
        if (line[0] == '[') continue;

        auto eq = line.find('=');
        if (eq == std::string::npos) continue;

        std::string key = line.substr(0, eq);
        auto key_start = key.find_first_not_of(" \t");
        auto key_end = key.find_last_not_of(" \t");
        if (key_start == std::string::npos) continue;
        key = key.substr(key_start, key_end - key_start + 1);

        std::string value = line.substr(eq + 1);
        auto val_start = value.find_first_not_of(" \t");
        if (val_start == std::string::npos) continue;
        value = value.substr(val_start);
        auto val_end = value.find_last_not_of(" \t");
        if (val_end != std::string::npos) {
            value = value.substr(0, val_end + 1);
        }
        if (value.size() >= 2 && value.front() == '"' && value.back() == '"') {
            value = value.substr(1, value.size() - 2);
        }

        if (key == "name") {
            name = value;
        } else if (key == "exec") {
            exec = value;
        } else if (key == "cover") {
            cover = value;
        }
    }

    if (!name || !exec) {
        return std::nullopt;
    }

    return GameEntry{
        .id = 0,
        .name = std::move(*name),
        .exec = std::move(*exec),
        .cover = std::move(cover),
    };
}
